src/Native/System.Security.Cryptography.Native/openssl.c

```diff
@@ -719,6 +719,224 @@ GetX509NameInfo(
 
 /*
 Function:
+CheckX509HostnameMatch
+
+Checks if a particular ASN1_STRING represents the entry in a certificate which would match against
+the requested hostname.
+
+Prameter sanRules: 0 for match rules against the subject CN, 1 for match rules against a SAN entry
+
+Return values:
+1 if the hostname is a match
+0 if the hostname is not a match
+Any negative number indicates an error in the arguments.
+*/
+static
+int
+CheckX509HostnameMatch(ASN1_STRING* candidate, const char* hostname, int cchHostname, char sanRules)
+{
+    assert(candidate);
+    assert(hostname);
+
+    if (!candidate->data || !candidate->length)
+    {
+        return 0;
+    }
+
+    // If the candidate is *.example.org then the smallest we would match is a.example.org, which is the same
+    // length. So anything longer than what we're matching against isn't valid.
+
+    // This assumption might not hold under IDNA, it might need to be more sophisticated
+    // "UTF character string length"
+    if (candidate->length > cchHostname)
+    {
+        return 0;
+    }
+
+    if (sanRules)
+    {
+        // RFC2818 says to use RFC2595 matching rules, but then gives an example that f*.com would match foo.com
+        // RFC2595 says that '*' may be used as the left name component, in which case it is a wildcard that does
+        // not match a '.'.
+        //
+        // In the interest of time, and the idea that it's better to err on the side of more restrictive,
+        // this implementation does not support mid-string wildcards.
+        //
+        // TODO (3444): Determine if we're too restrictive here.
+
+        char* candidateStr;
+        int i;
+        int hostnameFirstDot = -1;
+
+        // A GEN_DNS name is supposed to be a V_ASN1_IA5STRING.  If it isn't, we don't know how to read it.
+        if (candidate->type != V_ASN1_IA5STRING)
+        {
+            return 0;
+        }
+
+        // Great, candidateStr is just candidate->data!
+        candidateStr = (char*)candidate->data;
+
+        // First, verify that the string is alphanumeric, plus hyphens or periods and maybe starting with an asterisk.
+        for (i = 0; i < candidate->length; ++i)
+        {
+            char c = candidateStr[i];
+
+            if ((c < 'a' || c > 'z') &&
+                (c < '0' || c > '9') &&
+                (c != '.') &&
+                (c != '-') &&
+                (c != '*' || i != 0))
```

Comment:
I believe it's correct as is.  After checking the rest of the characters and finding it's thus far invalid, we check to see if it's a '*'.  If it's not, then the condition ```c != '*'``` is satisfied and we return 0 correctly as an error.  If it is a '*', then ```c != '*'``` is false and we proceed to check ```i != 0```... if it's the first character, then ```i != 0``` is false and we don't error out; if it's not the first character, then ```i != 0``` is true and we return 0 correctly.